src/bin/pg_autoctl/pgctl.c

```diff
@@ -918,6 +1030,100 @@ escape_recovery_conf_string(char *destination, const char *recoveryConfString)
 }
 
 
+/*
+ * prepare_primary_conninfo
+ */
+static int
+prepare_primary_conninfo(char *primaryConnInfo, int primaryConnInfoSize,
+						 const char *primaryHost, int primaryPort,
+						 const char *replicationUsername,
+						 const char *replicationPassword)
+{
+	char primaryConnInfoRaw[MAXCONNINFO] = { 0 };
+	char *connInfoEnd = primaryConnInfoRaw;
+
+ 	connInfoEnd += make_conninfo_field_str(connInfoEnd, "host", primaryHost);
+	connInfoEnd += make_conninfo_field_int(connInfoEnd, "port", primaryPort);
+	connInfoEnd +=
+		make_conninfo_field_str(connInfoEnd, "user", replicationUsername);
+
+	if (replicationPassword != NULL)
+	{
+		connInfoEnd +=
+			make_conninfo_field_str(connInfoEnd, "password", replicationPassword);
+	}
+
+	return escape_recovery_conf_string(primaryConnInfoRaw, primaryConnInfo);
+}
+
+
+/*
+ * pg_write_standby_signal writes the ${PGDATA}/standby.signal file that is in
+ * use starting with Postgres 12 for starting a standby server. The file only
+ * needs to exists, and the setup is to be found in the main Postgres
+ * configuration file.
+ */
+static bool
+pg_write_standby_signal(const char *configFilePath,
+						const char *pgdata,
+						const char *primaryConnInfo,
+						const char *replicationSlotName)
+{
+	GUC standby_settings[] = {
+		{ "primary_conninfo", (char  *)primaryConnInfo },
+		{ "primary_slot_name", (char  *) replicationSlotName},
+		{ "recovery_target_timeline", "latest"},
+		{ NULL, NULL }
+	};
+	char standbyConfigFilePath[MAXPGPATH];
+	char signalFilePath[MAXPGPATH];
+
+	log_trace("pg_write_standby_signal");
+
+	/*
+	 * First install the standby.signal file, so that if there's a problem
+	 * later and Postgres is started, it is started as a standby, with missing
+	 * configuration.
+	 */
+	join_path_components(signalFilePath, pgdata, "standby.signal");
+
+	log_info("Writing recovery configuration to \"%s\"", signalFilePath);
+
+	if (!write_file("", 0, signalFilePath))
+	{
+		/* write_file logs I/O error */
+		return false;
+	}
+
+	/*
+	 * Now write the standby settings to postgresql-auto-failover-standby.conf
+	 * and include that file from postgresql.conf.
+	 */
+	path_in_same_directory(configFilePath, AUTOCTL_STANDBY_CONF_FILENAME,
+						   standbyConfigFilePath);
+
+	/* we pass NULL as pgSetup because we now it won't be used... */
+	if (!ensure_default_settings_file_exists(standbyConfigFilePath,
+											 standby_settings,
+											 NULL))
+	{
+		return false;
+	}
+
+	if (!pg_include_config(standbyConfigFilePath,
+						   AUTOCTL_SB_CONF_INCLUDE_LINE,
+						   AUTOCTL_SB_CONF_INCLUDE_REGEX,
+						   AUTOCTL_CONF_INCLUDE_COMMENT))
+	{
+		log_error("Failed to prepare \"%s\" with standby settings",
```

Comment:
we have the signal file in place, but we can not include recovery config in postgresql.conf file. Is it okay to leave signal file.

Reply:
Added comments about why I think we should do things that way.